### PR TITLE
DRILL-8037: Add V2 JSON Format Plugin based on EVF

### PR DESCRIPTION
**Pull request by vdiravka**
# [DRILL-8037](https://issues.apache.org/jira/browse/DRILL-8037): Add V2 JSON Format Plugin based on EVF

## Description

This adds new V2 beta JSON Format Plugin based on the "Extended Vector Framework".
This is follow up [DRILL-6953](https://issues.apache.org/jira/browse/DRILL-6953) (was closed with the decision to merge it by small pieces).
So it is based on #1913 and [rev2](https://github.com/paul-rogers/drill/tree/DRILL-6953-rev2) work.

## Documentation
"V1" - "legacy" reader 
"V2" - new beta version JSON Format Plugin based on the result set loader. 
The V2 version is a bit more robust, and supports the row set framework. However, V1 supports unions and reading corrupted JSON files.

The new "V2" JSON scan is controlled by a new option:
`store.json.enable_v2_reader`, which is true by default in this PR.

Adds a "projection type" to the column writer so that the
JSON parser can receive a "hint" as to the expected type.
The hint is from the form of the projected column: a[0],
a.b or just a.
Therefore it supports schema provision. Example:
```
ALTER SESSION SET `store.json.enable_v2_reader` = true;
apache drill (dfs.tmp)> select * from test;
+---------------+-------+---+
|       a       |   e   | f |
+---------------+-------+---+
| {"b":1,"c":1} | false | 1 |
| {"b":1,"c":1} | null  | 2 |
| {"b":1,"c":1} | true  | 3 |
+---------------+-------+---+
apache drill (dfs.tmp)> create or replace schema (`e` BOOLEAN default 'false', `new` VARCHAR not null default 'schema evolution') for table test;
apache drill (dfs.tmp)> select * from test;
+-------+------------------+---------------+---+
|   e   |       new        |       a       | f |
+-------+------------------+---------------+---+
| false | schema evolution | {"b":1,"c":1} | 1 |
| null  | schema evolution | {"b":1,"c":1} | 2 |
| true  | schema evolution | {"b":1,"c":1} | 3 |
+-------+------------------+---------------+---+
```

## Testing
A lot of existing test cases are running for both readers. It is needed until V1 is still present in Drill code.

### CODE REVIEW

**Comment by vdiravka** [comment]
@paul-rogers I've cherry-picked commit from your [rev2](https://github.com/paul-rogers/drill/tree/DRILL-6953-rev2) branch. Didn't get all of them, because of big code conflicts and some of them are outdated.
Draft until all tests from [rev2](https://github.com/paul-rogers/drill/tree/DRILL-6953-rev2) are enabled.

**Comment by paul-rogers** [comment]
@vdiravka, thanks for doing this! It will take me a bit to remember where I left this. I'll do the review in small bits.

**Comment by paul-rogers** [comment]
@vdiravka, turns out there was one more commit that I had failed to push to `DRILL-6953-rev2`. Just did that. This commit contains tests for the extended types and for a provided schema. Please incorporate those additional changes so we have a complete set of work.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:350]: There are some changes here and above that need to be copied over.

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:33]: There is some cleanup in `TestComplexTypeWriter.java` to be copied across.

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:81]: This file has some extended type tests to be copied over.

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonEscapeAnyChar.java
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonEscapeAnyChar.java:122]: Changes in `TestJsonNanInf.java` need to be copied over.

**Review by paul-rogers — CHANGES_REQUESTED** [review]
Thanks again for merging this code across. I'm sure it was not an easy task.

I went though this PR, comparing it to a diff of my original branch with its base. Looks like you got most of the changes. I only found a few areas, noted below, where there were differences.

Please also grab that last commit I just pushed and I'll do a final review.

I saw that you had made a few good cleanups in doing the merge, so that means you probably took a good look at the code. I'll assume that counts as a review of my changes, and I'm just reviewing the merge (so that I'm not approving my own changes!)

**Review by vdiravka — COMMENTED** [review]
Hi @paul-rogers I am going to enable V2 by default. All tests passed and fixed, except cases with enabled options for UNION and JSON Corrupted files, but these features are experimental, so we are fine to go without them now.
My changes was in [branch](https://github.com/vdiravka/drill/commits/DRILL-8037_CI_rebase) and I have prepared them to be pushed for this PR in this [branch](https://github.com/vdiravka/drill/commits/DRILL-8037_final), just wait for CI and will do.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONFormatPlugin.java
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONFormatPlugin.java:93]: It is enabled by default. So what is temporary here?

File: exec/java-exec/src/main/resources/drill-module.conf
Comment (vdiravka) [reply exec/java-exec/src/main/resources/drill-module.conf:695]: yes, we want by default

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java
Comment (vdiravka) [reply exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:350]: What changes?

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (vdiravka) [reply exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:33]: Spaces? Could you point me what cleanup do you mean?

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (vdiravka) [reply exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:81]: Do you mean to add test cases for every extended type from the file?

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonEscapeAnyChar.java
Comment (vdiravka) [reply exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonEscapeAnyChar.java:122]: What changes?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONFormatPlugin.java
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONFormatPlugin.java:93]: Sorry, I saw this in the PR description:

> The new "V2" JSON scan is controlled by a new option:
> store.json.enable_v2_reader, which is false by default in this PR.

Which I thought meant that the V2 reader is not enabled by default, hence the suggested comment.

File: exec/java-exec/src/main/resources/drill-module.conf
Comment (paul-rogers) [reply exec/java-exec/src/main/resources/drill-module.conf:695]: If so, then please change the PR description quoted above.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java]: The code uses a method to do the alter session rather than building up a string. See [this file](https://github.com/paul-rogers/drill/blob/DRILL-6953-rev2/exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java#L283).

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:35]: See [this diff](https://github.com/paul-rogers/drill/compare/DRILL-6953-rev2...vdiravka:DRILL-8037#diff-343fdacd56be76f487ce9e5146b186dd3cfbd6fa3dce74c420941242605630f7), then search for `TestComplexTypeWriter`. You'll see some minor differences.

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestExtendedTypes.java:96]: See the diff above, search for `TestExtendedTypes`.

**Comment by paul-rogers** [comment]
Sorry, I perhaps read the diffs backward before: the things I thought were missing seem actually fine.

Just to double-check, there were some additional complex type handling and tests in [this commit](https://github.com/paul-rogers/drill/commit/7eefe01a623ade4ad387073686108ab658ce4322) that would be useful to add.

**Comment by vdiravka** [comment]
> Sorry, I perhaps read the diffs backward before: the things I thought were missing seem actually fine.
> 
> Just to double-check, there were some additional complex type handling and tests in [this commit](https://github.com/paul-rogers/drill/commit/7eefe01a623ade4ad387073686108ab658ce4322) that would be useful to add.

Hi Paul. I double checked and looks like the changes from commit is already incorporated. I made cheery-pick, resolved 3 merge conflicts and after that there are no new changes.

The other question what do you think is that useful to have dfs plugin config to switch to V1 JSON reader? I think, since the goal to switch to V2 fully, the system/session option is enough

**Comment by paul-rogers** [comment]
@vdiravka, the commit in question was something I pushed about a week ago: it was some last changes that were still on my local machine that I'd not yet pushed. Sorry about that! They were not on GitHub when you grabbed the original commits. For example, the new file `UtcDateValueListener.java` which I don't see in the list of files for this PR. Can you grab this commit also?

**Comment by vdiravka** [comment]
> @vdiravka, the commit in question was something I pushed about a week ago: it was some last changes that were still on my local machine that I'd not yet pushed. Sorry about that! They were not on GitHub when you grabbed the original commits. For example, the new file `UtcDateValueListener.java` which I don't see in the list of files for this PR. Can you grab this commit also?

For passing session option in tests I have created the new methods `enableSessionOption`, `disableSessionOption` based on the building the string `optionSettingQueriesForTestQuery` in `TestBuilder`.

I see, you've already commited this changes and file in `DRILL-7717: Support Mongo extended types in V2 JSON loader` #2068
Screenshot from 2021-12-04 14-53-08

**Review by paul-rogers — APPROVED** [review]
@vdiravka, thanks for the explanations. Hard to remember stuff that happened over a year ago...

The changes look good. Thanks much for getting this merged!

LGTM. +1

**Comment by vdiravka** [comment]
I found one regression for LATTERAL and UNNEST operators with V2. SchemaChange starts to be happened, but it shouldn't.
`TestE2EUnnestAndLateral#testMultipleBatchesLateral_WithGroupByInParent` shows it after reverting changes in `HashAggTemplate`. Working on fix now

**Comment by luocooong** [comment]
@vdiravka Will the fix included in this PR or a new separate PR?

**Comment by vdiravka** [comment]
I'll try to prepare the fix today in this PR. If it will take longer, than it is fine to include in the separate PR

**Comment by paul-rogers** [comment]
@vdiravka reached out to me on this bug. His explanation:

> The issue is the schema is changed for the second batch and it is reported by SchemaTracker#isSameSchema
I suppose V1 reader just compared the same field type, but the new one compares vectors by identity. So schemaChange is right thing for that test case. But Drill Hash aggregate doesn't support schema change.

Here is my analysis:

You are tight that the `SchemaTracker` is the thing in EVF which looks for schema changes. The rule is pretty simple: if the the kind of vector differs from the previous batch, then a schema change occurred. This is based, in part, on the observation that the sort operator can't handle any schema changes at all: not INT/BIGINT, not NOT NULL to NULL. Sort combines vectors, and if they are of a different type, the values won't fit together. The same is probably true of the hash agg.

OK, so we agree that a type change is bad. What you're describing is the same type, but a different vector. This is a bug in several ways, and not the way you might think.

First, it seems logical that if column x is a non-null INT in one batch, and is a non-null INT in the next batch, that no schema change has occurred. But, Drill will surprise you. Most operators assume that the vector itself has not changed: only the data stored within the vector. This is a subtle point, so let me expand it.

A vector is a permanent holder or a column. The vector holds one or more data buffers. Those buffers change from batch to batch. Think of it as a bucket brigade: the old-style way that people fought fires. A chain of people forms (the vectors). They hand buckets one to the next (the data buffers).

Now, I was pretty surprised when I first discovered this. So was Boaz when he wrote hash agg. But, most of Drill's code assumes that, in setup, the code binds to a specific vector. In the next batch, that same binding is valid; only the data changes. This shows up in code gen and other places. In fact, a goodly part of EVF is concerned with this need for "vector continuity" even across readers. (Let that sink in: two readers that have nothing to do with each other must somehow still share common vectors. That alone should tell us the vector design in Drill is a bit strange and needs rethinking. But, that's another topic.)

So, now let's review the bug. On the one hand, `SchemaTracker` has noticed the vectors changed, and is (correctly) telling the downstream operators that they must redo their bindings to the new vector.

But, on the other hand, `SchemaTracker` has been presented with two different vectors for the same column. That should never happen unless there is an actual type change (non-null to null, INT to BIGINT, etc.) There are elaborate mechanisms to reuse vectors from one reader to the next.

So, the first thing to check is if the types are at all different. (If the "major types" differ.) If so, then we have a legitimate schema change that Drill can't handle.

Now, if you find that the types are the same, then we have a bug in EVF, specifically in the `ResultVectorCacheImpl` class, since that's the thing which is supposed to do the magic.

Let me ask another question. When this occurs, is it in the first batch of the second (or later) reader? If so, then that reader should have pulled the vector from the cache. The question is, why didn't it?

If the error occurs within the first reader, then the bug is more subtle. How could the JSON reader accomplish that? The `ResultSetLoader` hides all that detail, and it defers to the `ResultVectorCacheImpl `for vector continuity.

Yet another question is why these tests didn't fail 3 months ago when the tests first ran. And, why did the tests pass way back when I wrote this code? Did anything change elsewhere that could trigger this? I don't see how, but we should probably ask the question. 

Fortunately, these are unit tests, so we should be able to sort out the issues without too much trouble.

I'll also note that there were a few fixes to EVF V2 that I made in my PR, but those were around a subtle but in implicit columns.

Suggestion: give the above a shot to see if you can see the problem. Otherwise, tomorrow I'll try to squeeze in time to grab this branch and do some debugging. After all, I wrote this stuff originally, so I should try to make it work.

**Comment by paul-rogers** [comment]
Oddly, `TestE2EUnnestAndLateral` works just fine in Eclipse on my machine with this branch. Is that the correct test which was failing? Also, I see in the panel here in GitHub that all tests passed. Can you explain the issue a bit more?

**Comment by vdiravka** [comment]
Hi @paul-rogers I have rebased the branch to master branch. And in separate new commit removed the hack, which hid the schema change in the `HashAggTemplate` (and actually one row is missing in query result, just actually test case doesn't check it).
Thanks for explanation how vectors is working, it helped me. It is clear now, that schema is changing due to `RepeatedMapVector` [can't be obtained from the cache](https://github.com/apache/drill/blob/317f164791bbbe8f937eb452b49e92c34f1c0333/exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/impl/ColumnBuilder.java#L220):
```
      // Don't get the map vector from the vector cache. Map vectors may
      // have content that varies from batch to batch. Only the leaf
      // vectors can be cached.
```
Obtaining vector from cache here leads to errors in this and other test cases:
`mapVector = (RepeatedMapVector) parent.vectorCache().vectorFor(mapColSchema.schema());`

```
org.apache.drill.common.exceptions.UserRemoteException: EXECUTION_ERROR ERROR: null

Read failed for reader: JsonBatchReader
....
Caused by: java.lang.AssertionError: 
	at org.apache.drill.exec.physical.resultSet.impl.TupleState$MapState.addOutputColumn(TupleState.java:475)
	at org.apache.drill.exec.physical.resultSet.impl.ColumnState.buildOutput(ColumnState.java:321)
	at org.apache.drill.exec.physical.resultSet.impl.TupleState.updateOutput(TupleState.java:206)
	at org.apache.drill.exec.physical.resultSet.impl.TupleState.updateOutput(TupleState.java:217)
	at org.apache.drill.exec.physical.resultSet.impl.TupleState$RowState.updateOutput(TupleState.java:430)
	at org.apache.drill.exec.physical.resultSet.impl.ResultSetLoaderImpl.harvest(ResultSetLoaderImpl.java:716)
```
So as for me looks like we need to implement supporting schema change for hashAgg operator or obtaining `RepeatedMapVector` from the cache. I lean towards the latter. What do you think?

**Comment by paul-rogers** [comment]
@vdiravka, good sleuthing! You did indeed find the hole in the system. Map (and repeated map) vectors are special: they are just holders for the actual data vectors. If they are reused, we get all the previous map members, which may or may not be a problem. I guess it would be a problem if reader 1 has a.b be an INT, while reader 2 wants a.b to be a VARCHAR.

I guess a question is whether the HashAgg maintains a pointer to the map itself, or only the physical columns within it. It has no pointers to the map itself, we can special-case maps: they are considered the same schema if their contents are the same, whether or not the map vector itself is the same.

Another choice would be to store the map vector in the cache, but strip all the physical columns out of it when the caller asks for it again. The caller then reassembles the physical columns, also from the cache, and hopefully creates the same map structure as the previous reader.

Based on what you learned of the HashAgg, which of these might work?

**Comment by paul-rogers** [comment]
@vdiravka, which test do I run to see the failure? I tried that one test mentioned above, but it worked for me.

Actually, I should probably create a separate unit test for this scenario. I'll do that to see if I can reproduce the issue. If so, then I'll see if I can find a fix, perhaps one of the ideas mentioned above.

Then, we can validate the fix on that test case you have which is failing.

**Comment by vdiravka** [comment]
@paul-rogers 
I have rebased this PR with a mater branch and fixed [DRILL-8195](https://issues.apache.org/jira/browse/DRILL-8195). So here are no new changes after last review (except "rebase to master" commit and DRILL-8195. "Rebase to master" commit will be squashed to previous one after review)

Additionally [DRILL-8196](https://issues.apache.org/jira/browse/DRILL-8196), [DRILL-8197](https://issues.apache.org/jira/browse/DRILL-8197) and [DRILL-8201](https://issues.apache.org/jira/browse/DRILL-8201) are opened and going to be resolved in separate PRs soon.

**Review by paul-rogers — APPROVED** [review]
@vdiravka, thanks again for merging the code.
LGTM +1

**Comment by vdiravka** [comment]
Thanks for the review @paul-rogers 
Fixup(squash) the "Rebase to the master branch" commit now and will merge the PR after CI passing

**Pull request merged**
